### PR TITLE
ADR Agents - Assistant V2 Support

### PR DESCRIPTION
**Pull request by crickman**
### Motivation and Context

Open AI has release the _Assistants V2_ API.  This builds on top of the V1 _assistant_ concept, but also invalidates certain V1 features.  In addition, the _dotnet_ API that supports _Assistant V2_ features is entirely divergent on the `Azure.AI.OpenAI.Assistants` SDK that is currently in use.

### Description

ADR describes changes to `Microsoft.SemanticKernel.Agents.OpenAI` project / package.

Changes prototyped in this PR: https://github.com/microsoft/semantic-kernel/pull/7126

PR completion is gated by the V2 connector migration work due to versioning constraints.

> ADR Reviewed on July 18th, 2024

### Contribution Checklist

- [X] The code builds clean without any errors or warnings
- [X] The PR follows the [SK Contribution Guidelines](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md) and the [pre-submission formatting script](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md#development-scripts) raises no violations
- [X] All unit tests pass, and I have added new tests where possible
- [X] I didn't break anyone :smile:

### CODE REVIEW

File: docs/decisions/0049-agents-assistantsV2.md [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+### Configuration Classes
+
+Specific configuration/setttings classes are introduced to support the ability to define assistant behavior at each of the supported articulation points (i.e. _assistant_, _thread_, & _run_).
+
+
+#### Assistant Definition
+
+The `OpenAIAssistantDefinition` was previously used only when enumerating a list of stored agents.  It has been evolved to also be used as input for creating and agent and exposed as a discrete property on the `OpenAIAssistantAgent` instance.
+
+This includes optional _execution_settings_ for defining default _run_ behavior.  Since these execution settings are not part of the remote assistant definition, they are persisted in the assistant metadata for when an existing agent is retrieved.
```
Comment (moonbox3) [inline docs/decisions/0049-agents-assistantsV2.md:29]: Is this a subclass of the current `PromptExecutionSettings`? So one has the ability to control the `FunctionChoiceBehavior`?

File: docs/decisions/0049-agents-assistantsV2.md [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+When invoking an `OpenAIAssistantAgent` directly (no-chat), settings that only apply to a discrete run may be specified.  These settings are defined as `OpenAIAssistantInvocationSettings`.
+
+<p align="center">
+<kbd><img src="diagrams/assistant-invocationsettings.png" style="width: 220pt;"></kbd>
+</p>
+
+
+#### Thread Creation Settings
+
+When invoking an `OpenAIAssistantAgent` directly (no-chat), a thread must be explicitly managed.  When doing so, thread specific settings may be specified.  These settings are defined as `OpenAIThreadCreationSettings`.
```
Comment (moonbox3) [inline docs/decisions/0049-agents-assistantsV2.md:47]: Is it necessary to have `EnableCodeInterpreter` in both these settings and `OpenAIAssistantInvocationSettings`? Which would have precedence?

File: docs/decisions/0049-agents-assistantsV2.md [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+**Invoke**|Invoke the assistant agent (no chat)
+**GetChannelKeys**|Inherited from `Agent`
+**CreateChannel**|Inherited from `Agent`
+
+
+### Vector Store Support
+
+_Vector Store_ support is required in order to enable usage of the `file-search` tool.  
+
+In alignment with V2 treaming of the `FileClient`, the caller may also directly target `VectorStoreClient` from the _OpenAI SDK_.
```
Comment (moonbox3) [inline docs/decisions/0049-agents-assistantsV2.md:97]:
```suggestion
In alignment with V2 streaming of the `FileClient`, the caller may also directly target `VectorStoreClient` from the _OpenAI SDK_.
```

File: docs/decisions/0049-agents-assistantsV2.md [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+|Class Name|Description|
+---|---
+**OpenAIAssistantAgent**|An `Agent` based on the Open AI Assistant API
+**OpenAIAssistantChannel**|An 'AgentChannel' for `OpenAIAssistantAgent` (associated with a _thread-id_.)
+**OpenAIAssistantDefinition**|All of the metadata / definition for an Open AI Assistant.  Unable to use the _Open AI API_ model due to implementation constraints (constructor not public).
+**OpenAIAssistantExecutionSettings**|Setting that affect the _run_, but defined globally for the agent/assistant.
+**OpenAIAssistantInvocationSettings**|Settings bound to a discrete run, used for direct (no chat) invocation.
+**OpenAIServiceConfiguration**|Describes the service connection and used to create the `OpenAIClient`
+
+### Run Processing
```
Comment (moonbox3) [inline docs/decisions/0049-agents-assistantsV2.md:112]: I love this section and the high level overview about what it takes to run the OpenAI Assistant agent. One thing I noticed as I look into the Python OpenAI assistants V2 library is that it is mainly built to accept keyword-arguments, and not specific classes like `RunCreationOptions` or other specific object used in the dotnet SDK. You can view the OpenAI doc [here](https://platform.openai.com/docs/api-reference/runs/createRun) or the AzureOpenAI doc [here](https://learn.microsoft.com/en-us/azure/ai-services/openai/assistants-reference-runs?tabs=python), as an example.

File: docs/decisions/0049-agents-assistantsV2.md [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+
+
+### Vector Store Support
+
+_Vector Store_ support is required in order to enable usage of the `file-search` tool.  
+
+In alignment with V2 treaming of the `FileClient`, the caller may also directly target `VectorStoreClient` from the _OpenAI SDK_.
+
+
+### Class Inventory
```
Comment (moonbox3) [inline docs/decisions/0049-agents-assistantsV2.md:100]: All of the classes mentioned here feels like a lot of different config classes to understand and manage -- there is the definition, the execution settings, the invocation settings, and the service configuration. It feels a bit overwhelming, at first glance.

File: docs/decisions/0049-agents-assistantsV2.md (hunk 1 above)
Comment (crickman) [reply docs/decisions/0049-agents-assistantsV2.md:29]: No, its not.  It corresponds to the core assistant model:  https://platform.openai.com/docs/api-reference/assistants/object

The behavioral settings are defined by the associated `When invoking an `OpenAIAssistantAgent` directly (no-chat), settings that only apply to a discrete run may be specified.  These settings are defined as `OpenAIAssistantInvocationSettings`: `OpenAIAssistantDefinition.ExecutionSettings`.

> Note: I'm working on dialing in the function-choice / tool-choice option.

File: docs/decisions/0049-agents-assistantsV2.md (hunk 2 above)
Comment (crickman) [reply docs/decisions/0049-agents-assistantsV2.md:47]: The thread-settings have precedence over the assistant settings (when specified).

The invokcation-settings have precedence over all setting (when specified).  

Funny, it just occured to me to clarify this in the ADR...good call-out.

File: docs/decisions/0049-agents-assistantsV2.md (hunk 4 above)
Comment (crickman) [reply docs/decisions/0049-agents-assistantsV2.md:112]: Yes, this is an internal implementation detail of the .Net SDK (not visible to the caller).  The underlying service endpoint / network traffic should be identical and logically equivalent.

File: docs/decisions/0049-agents-assistantsV2.md (hunk 5 above)
Comment (crickman) [reply docs/decisions/0049-agents-assistantsV2.md:100]: These should all map directly to the core API:

https://platform.openai.com/docs/api-reference/assistants

I'm just mirroring the V2 behavior.  Open to suggestions if you have something in mind.

File: docs/decisions/0049-agents-assistantsV2.md (hunk 1 above)
Comment (crickman) [reply docs/decisions/0049-agents-assistantsV2.md:29]: Looking a little deeper, I wonder if `ToolCallBehavior` isn't the class that might be extended here?

File: docs/decisions/0049-agents-assistantsV2.md (hunk 5 above)
Comment (crickman) [reply docs/decisions/0049-agents-assistantsV2.md:110]: Definition => Assistant
Thread => Thread
Execution and Invocation => Run
Service => Access

File: docs/decisions/0049-agents-assistantsV2.md [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+
+_Vector Store_ support is required in order to enable usage of the `file-search` tool.  
+
+In alignment with V2 streaming of the `FileClient`, the caller may also directly target `VectorStoreClient` from the _OpenAI SDK_.
+
+
+### Class Inventory
+This section provides an overview / inventory of all the public surface area described in this ADR.
+
+|Class Name|Description|
```
Comment (moonbox3) [inline docs/decisions/0049-agents-assistantsV2.md:129]: Is this table missing the `OpenAIThreadCreationSettings` class?

Comment (crickman) [reply docs/decisions/0049-agents-assistantsV2.md:57]: Indeed, updated.

**Review by moonbox3 — APPROVED** [review]

**Review by alliscode — APPROVED** [review]

**Pull request merged**